<?xml version="1.0"?>
<launch>

    <arg name="model" default="$(find steer_bot_description)/urdf/steer_bot.urdf.xacro"/>
    <arg name="namespace" default="/steer_bot" />
    <arg name="world_name" default="worlds/empty_world"/>
    <arg name="cmd_timeout" default="0.5"/>
    <arg name="x" default="0.0"/>
    <arg name="y" default="0.0"/>
    <arg name="z" default="1.0" />
    <arg name="roll" default="0.0"/>
    <arg name="pitch" default="0.0"/>
    <arg name="yaw" default="0.0"/>

    <group ns="$(arg namespace)">

        <!-- Gazebo  -->
        <include file="$(find gazebo_ros)/launch/empty_world.launch">
            <arg name="world_name" value="$(find Autonomous_Systems_Project_Team_8)/worlds/Racing_Track.world"/>
        </include>

        <!-- Load the robot description -->
        <param name="robot_description" command="$(find xacro)/xacro $(arg model)"/>

        <!-- Load ros_controllers configuration parameters -->
        <rosparam file="$(find steer_bot_control)/config/ctrl_ackermann_steering_controller.yaml" command="load" ns="$(arg namespace)" />
        <rosparam file="$(find steer_bot_control)/config/ctrl_gains.yaml" command="load" ns="$(arg namespace)" />
        <rosparam file="$(find steer_bot_control)/config/ctrl_joint_state_publisher.yaml" command="load" ns="$(arg namespace)" />
        <rosparam file="$(find steer_bot_control)/config/ctrl_steer_bot_hardware_gazebo.yaml" command="load" ns="$(arg namespace)" />

        <!-- Spawn the controllers -->
        <node pkg="controller_manager" type="spawner" name="controller_spawner" ns="$(arg namespace)"
            args="joint_state_publisher ackermann_steering_controller"
             respawn="false" />
        
        <!-- Launch  the robot state publisher -->
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
            <param name="publish_frequency" value="50.0"/>
        </node>

        <!-- Spawn robot in Gazebo -->
  
        <node name="spawn_vehicle" pkg="gazebo_ros" type="spawn_model"
            args="-urdf -param robot_description -model steer_bot
                -gazebo_namespace /$(arg namespace)/gazebo
                -x $(arg x) -y $(arg y) -z $(arg z)
                -R $(arg roll) -P $(arg pitch) -Y $(arg yaw)"
            respawn="false" output="screen" />

    </group>

    <group ns="$(arg namespace)">
       <node pkg="Autonomous_Systems_Project_Team_8" name="Control_Module_Lateral_Motion_Ms_5" type="Control_Module_Lateral_Motion_Ms_5.py" output="screen">
          <param name="Wheel_Base" value="0.38"/> 
          <param name="Lateral_Controller" value="Pure_Pursuit"/> 
          <remap from="/gazebo/model_states" to="$(arg namespace)/gazebo/model_states"/>
          <remap from="/Control_Action_Steering" to="$(arg namespace)/Control_Action_Steering"/>
          <remap from="/Longitudinal_Driving_Velocity" to="$(arg namespace)/Longitudinal_Driving_Velocity"/>
          <remap from="/Lateral_Distance" to="$(arg namespace)/Lateral_Distance"/>
       </node>
       
       <node pkg="Autonomous_Systems_Project_Team_8" name="Control_Module_Longitudinal_Motion_Ms_5" type="Control_Module_Longitudinal_Motion_Ms_5.py" output="screen">
          <param name="Wheel_Base" value="0.38"/> 
          <remap from="/gazebo/model_states" to="$(arg namespace)/gazebo/model_states"/>
          <remap from="/Control_Action_Driving_Velocity" to="$(arg namespace)/Control_Action_Driving_Velocity"/>
          <remap from="/Longitudinal_Driving_Velocity" to="$(arg namespace)/Longitudinal_Driving_Velocity"/>
          <remap from="/Lateral_Distance" to="$(arg namespace)/Lateral_Distance"/>
       </node>
       
       <node pkg="Autonomous_Systems_Project_Team_8" name="Steer_Bot_Control_Module_Ms_5" type="Steer_Bot_Control_Module_Ms_5.py" output="screen">
          <remap from="/Control_Module_Output" to="$(arg namespace)/ackermann_steering_controller/cmd_vel"/>
          <remap from="/Control_Action_Steering" to="$(arg namespace)/Control_Action_Steering"/>
          <remap from="/Control_Action_Driving_Velocity" to="$(arg namespace)/Control_Action_Driving_Velocity"/>
       </node>
       
       <node pkg="Autonomous_Systems_Project_Team_8" name="Autonomous_Systems_MS_5_Planning_Team_8" type="Autonomous_Systems_MS_5_Planning_Team_8.py" output="screen">
          <remap from="/Longitudinal_Driving_Velocity" to="$(arg namespace)/Longitudinal_Driving_Velocity"/>
          <remap from="/Lateral_Distance" to="$(arg namespace)/Lateral_Distance"/>
          <remap from="/gazebo/model_states" to="$(arg namespace)/gazebo/model_states"/>
       </node>
       
       <node pkg="Autonomous_Systems_Project_Team_8" name="Autonomous_Systems_MS_5_Localization_Team_8" type="Autonomous_Systems_MS_5_Localization_Team_8.py" output="screen">
       </node>
       
    </group>

    <node pkg="rqt_graph" name="rqt_graph" type="rqt_graph" />

</launch>
